src/permute.js

```diff
@@ -1,3 +1,45 @@
-export default function(source, keys) {
-  return Array.from(keys, key => source[key]);
+export default function (source, keys) {
+  if (keys === undefined) {
+    return permute(source);
+  } else {
+    return Array.from(keys, (key) => source[key]);
+  }
+}
+
+// Based on https://github.com/google/guava/blob/f4b3f611c4e49ecaded58dcb49262f55e56a3322/guava/src/com/google/common/collect/Collections2.java#L622-L683
```

Comment:
Maybe this link can be moved into the docs?

Reply:
Something like.

This implementation was inspired by the [permute algorithm in Guava Collections](https://github.com/google/guava/blob/f4b3f611c4e49ecaded58dcb49262f55e56a3322/guava/src/com/google/common/collect/Collections2.java#L622-L683).